packages/libs/error-stack/src/ext/iter.rs

```diff
@@ -0,0 +1,353 @@
+where
+    I: Iterator,
+    I::Item: ResultExt,
+    A: Display + Debug + Send + Sync + 'static + Clone,
+{
+    type Item = I::Item;
+
+    #[track_caller]
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.iterator.next() {
+            Some(item) => Some(item.attach_printable(self.context.clone())),
+            None => None,
+        }
+    }
+}
+
+pub struct IteratorWithContext<I, A> {
+    iterator: I,
+    context: A,
+}
+
+impl<I, A> Iterator for IteratorWithContext<I, A>
+where
+    I: Iterator,
+    I::Item: ResultExt,
+    A: Context + Clone,
+{
+    type Item = Result<<I::Item as ResultExt>::Ok, Report<A>>;
+
+    #[track_caller]
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.iterator.next() {
+            Some(item) => Some(item.change_context(self.context.clone())),
+            None => None,
+        }
+    }
+}
+
+pub struct IteratorWithLazyContext<I, A> {
+    iterator: I,
+    context: A,
+}
+
+impl<I, F, A> Iterator for IteratorWithLazyContext<I, F>
+where
+    I: Iterator,
+    I::Item: ResultExt,
+    F: FnOnce() -> A + Clone,
+    A: Context,
+{
+    type Item = Result<<I::Item as ResultExt>::Ok, Report<A>>;
+
+    #[track_caller]
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.iterator.next() {
+            Some(item) => Some(item.change_context_lazy(self.context.clone())),
+            None => None,
+        }
+    }
+}
+
+pub struct IteratorWithLazyPrintableAttachement<I, A> {
+    iterator: I,
+    context: A,
+}
+
+impl<I, F, A> Iterator for IteratorWithLazyPrintableAttachement<I, F>
+where
+    I: Iterator,
+    I::Item: ResultExt,
+    F: FnOnce() -> A + Clone,
+    A: Context,
+{
+    type Item = I::Item;
+
+    #[track_caller]
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.iterator.next() {
+            Some(item) => Some(item.attach_printable_lazy(self.context.clone())),
+            None => None,
+        }
+    }
+}
+
+#[cfg(test)]
+/// Just tests that some simple uses of `IteratorExt` (and related types) works.
+mod test_simple_compiles {
+    use super::*;
+    use crate::test_helper::messages;
+
+    #[derive(Debug)]
+    pub struct UhOhError;
+
+    impl core::fmt::Display for UhOhError {
+        fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+            f.write_str("UhOhError")
+        }
+    }
+
+    impl crate::Context for UhOhError {}
+
+    #[test]
+    fn report_with_vec() {
+        use alloc::{string::ToString, vec};
+
+        let items = vec![
+            Ok(1),
+            Ok(1),
+            Ok(2),
+            Ok(5),
+            Ok(14),
+            Ok(42),
+            Ok(132),
+            Ok(429),
+            Ok(1430),
+            Ok(4862),
+            Ok(16796),
+            Err(UhOhError),
+        ]
+        .into_iter()
+        .map(crate::ext::result::IntoReport::into_report);
+
+        let report = items.into_iter().attach("context".to_string());
```

Comment:
Let's use `attach_lazy` here as we calling a function (similar to `unwrap_or` vs `unwrap_or_else`):
```suggestion
        let report = items.into_iter().attach_lazy("context".to_string());
```